File: app/src/main/java/forpdateam/ru/forpda/realm/DbMigration.java
package forpdateam.ru.forpda.realm;

import android.support.annotation.NonNull;
import android.util.Log;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import io.realm.DynamicRealm;
import io.realm.RealmMigration;
import io.realm.RealmObjectSchema;
import io.realm.RealmSchema;

/**
 * Created by isanechek on 29.08.16.
 * <p>
 * Это хрень нужна для того чтобы мигрировать с одной версии обьекта на другой.
 * Ну вдруг там поля поменялись или еще чего.
 */

public class DbMigration implements RealmMigration {

    @Override
    public void migrate(@NonNull DynamicRealm realm, long oldVersion, long newVersion) {
        RealmSchema schema = realm.getSchema();

        /*
            for oldest versions
            */
        RealmObjectSchema oldFavSchema = schema.get("FavItemBd");
        if (oldFavSchema != null && !oldFavSchema.hasField("isForum")) {
            oldFavSchema.addField("isForum", boolean.class);
        }
        RealmObjectSchema oldHistorySchema = schema.get("HistoryItemBd");
        if (oldHistorySchema != null && !oldHistorySchema.hasField("url")) {
            oldHistorySchema.addField("url", String.class);
        }

        if (oldVersion == 1) {
            RealmObjectSchema favSchema = schema.get("FavItemBd");
            if (favSchema != null) {
                favSchema
                        .removeField("isNewMessages")
                        .removeField("info")
                        .addField("isNew", boolean.class)
                        .addField("isPoll", boolean.class)
                        .addField("isClosed", boolean.class);
            }

            oldVersion++;
        }

        if (oldVersion == 2) {
            RealmObjectSchema historySchema = schema.get("HistoryItemBd");
            if (historySchema != null) {
                SimpleDateFormat oldDateFormat = new SimpleDateFormat("MM.dd.yy, HH:mm", Locale.getDefault());
                SimpleDateFormat newDateFormat = new SimpleDateFormat("dd.MM.yy, HH:mm", Locale.getDefault());
                historySchema
                        .transform(dynamicRealmObject -> {
                            String dateString = dynamicRealmObject.getString("date");
                            Date date = new Date();
                            try {
                                date = oldDateFormat.parse(dateString);
                            } catch (ParseException e) {
                                e.printStackTrace();
                            }
                            Log.d("SUKA", "DATES " + dateString + " : " + newDateFormat.format(date));
                            dynamicRealmObject.setString("date", newDateFormat.format(date));
                        });
            }

            oldVersion++;
        }
    }
}

File: app/src/main/java/forpdateam/ru/forpda/common/realm/DbMigration.kt
package forpdateam.ru.forpda.common.realm

import io.realm.kotlin.dynamic.getNullableValue
import io.realm.kotlin.migration.AutomaticSchemaMigration
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Created by isanechek on 29.08.16.
 *
 *
 * Это хрень нужна для того чтобы мигрировать с одной версии обьекта на другой.
 * Ну вдруг там поля поменялись или еще чего.
 */
class DbMigration : AutomaticSchemaMigration {

    override fun migrate(migrationContext: AutomaticSchemaMigration.MigrationContext) {
        var updVersion = migrationContext.oldRealm.version().version

        migrateOldest(migrationContext)

        if (updVersion == 1L) {
            migrateV2(migrationContext)
            updVersion++
        }

        if (updVersion == 2L) {
            migrateV3(migrationContext)
            updVersion++
        }

        if (updVersion == 3L) {
            migrateV4(migrationContext)
            updVersion++
        }
    }

    private fun migrateOldest(context: AutomaticSchemaMigration.MigrationContext) {
        context.enumerate("FavItemBd") { old, new ->
            if (old.getObject("isForum") == null) {
                new?.set("isForum", false)
            }
        }
        context.enumerate("HistoryItemBd") { old, new ->
            if (old.getObject("url") == null) {
                new?.set("url", false)
            }
        }
    }

    private fun migrateV2(context: AutomaticSchemaMigration.MigrationContext) {
        context.enumerate("FavItemBd") { old, new ->
            // removed isNewMessages
            // removed info
            new?.set("isNew", false)
            new?.set("isPoll", false)
            new?.set("isClosed", false)
        }
    }

    private fun migrateV3(context: AutomaticSchemaMigration.MigrationContext) {
        val oldDateFormat = SimpleDateFormat("MM.dd.yy, HH:mm", Locale.getDefault())
        val newDateFormat = SimpleDateFormat("dd.MM.yy, HH:mm", Locale.getDefault())
        context.enumerate("HistoryItemBd") { old, new ->
            val newDate = old.getNullableValue<String>("date")?.also {
                var date = Date()
                try {
                    date = requireNotNull(oldDateFormat.parse(it))
                } catch (e: ParseException) {
                    e.printStackTrace()
                }
                newDateFormat.format(date)
            }
            new?.set("date", newDate)
        }
    }

    private fun migrateV4(context: AutomaticSchemaMigration.MigrationContext) {
        context.enumerate("FavItemBd") { old, new ->
            val nullString: String? = null
            new?.set("curatorId", 0)
            new?.set("curatorNick", nullString)
            new?.set("subType", nullString)
        }
    }
}
